locator/abstract_replication_strategy.cc

```diff
@@ -294,8 +294,16 @@ abstract_replication_strategy::get_pending_address_ranges(const token_metadata_p
     token_metadata temp;
     temp = co_await tmptr->clone_only_token_map();
     co_await temp.update_normal_tokens(pending_tokens, pending_address);
-    for (auto& x : co_await get_address_ranges(temp, pending_address)) {
-            ret.push_back(x.second);
+    for (const auto& t : temp.sorted_tokens()) {
+        auto eps = co_await calculate_natural_endpoints(t, temp);
+        for (auto ep : eps) {
+            if (ep == pending_address) {
+                dht::token_range_vector r = temp.get_primary_ranges_for(t);
+                rslogger.debug("get_pending_address_ranges: token={} primary_range={} endpoint={}", t, r, pending_address);
+                std::copy(r.begin(), r.end(), std::back_inserter(ret));
+                break;
+            }
```

Comment:
we append the result of `get_primary_ranges_for(t)` to `ret` only if `pending_address` is contained in `calculate_natural_endpoints(t, temp);`

Reply:
Re-looking at it, I think we better move the implementation into the effective replication map since there there is no need to calculate_natural_endpoints for every token. It's already pre-calculated in the replication map.

Reply:
Ah. Can we write it as `if (eps.contains(pending_address)`? It's much clearer.

Reply:
In fact I started working today on a change to caculate_natural_endpoints to return a endpoint_set which is a utils::sequenced_set, i.e. a set/vector hybrid which can be used both ways.  It is built internally in network_topology_replication_strategy anyhow today.